server/bin/pbench-index.py

```diff
@@ -181,393 +188,15 @@ def main(options, name):
-
-            report_fname = Path(tmpdir, f"{name}.{idxctx.TS}.report")
-            with report_fname.open(mode="w") as fp:
-                print(subj, file=fp)
-                if idx > 0:
-                    print("\nIndexed Results\n===============", file=fp)
-                    with indexed.open() as ifp:
-                        for line in sorted(ifp):
-                            print(line.strip(), file=fp)
-                if err > 0:
-                    print(
-                        "\nResults producing errors" "\n========================",
-                        file=fp,
-                    )
-                    with erred.open() as efp:
-                        for line in sorted(efp):
-                            print(line.strip(), file=fp)
-                if skp > 0:
-                    print("\nSkipped Results\n===============", file=fp)
-                    with skipped.open() as sfp:
-                        for line in sorted(sfp):
-                            print(line.strip(), file=fp)
-            try:
-                report.post_status(tstos(idxctx.time()), "status", report_fname)
-            except Exception:
-                pass
-
-    return res
-
-
-###########################################################################
-# Options handling
 if __name__ == "__main__":
-    run_name = Path(sys.argv[0]).name
-    run_name = run_name if run_name[-3:] != ".py" else run_name[:-3]
-    if run_name not in ("pbench-index",):
-        print(f"unexpected command file name: {run_name}", file=sys.stderr)
-        sys.exit(1)
-    parser = ArgumentParser(
-        f"Usage: {run_name} [--config <path-to-config-file>] [--dump-index-patterns]"
-        " [--dump_templates]"
-    )
-    parser.add_argument(
-        "-C",
-        "--config",
-        dest="cfg_name",
-        default=os.environ.get("_PBENCH_SERVER_CONFIG"),
-        help="Specify config file",
-    )
-    parser.add_argument(
-        "-I",
-        "--dump-index-patterns",
-        action="store_true",
-        dest="dump_index_patterns",
-        default=False,
-        help="Emit a list of index patterns used",
-    )
-    parser.add_argument(
-        "-Q",
-        "--dump-templates",
-        action="store_true",
-        dest="dump_templates",
-        default=False,
-        help="Emit the full JSON document for each index template used",
-    )
-    parser.add_argument(
-        "-T",
-        "--tool-data",
-        action="store_true",
-        dest="index_tool_data",
-        default=False,
-        help="Only index tool data, assumes run data already exists",
-    )
-    parser.add_argument(
-        "-R",
-        "--re-index",
-        action="store_true",
-        dest="re_index",
-        default=False,
-        help="Perform re-indexing of previously indexed data",
-    )
-    parsed = parser.parse_args()
-    status = main(parsed, run_name)
+    status = main()
```

Comment:
Originally, `main()` was a callable method that was not attached to how it got command line parameters.  I had `options` and `run_name` arguments.  It was done this way so that `main()` could be called without outside of a CLI context.

Once you pull the command line parsing code out of it, and place it back at the runtime module level, there is not much left to `main()`.  The behaviors of how tar balls are gathered for indexing are embedded in the new `Index` class.  While this moves code around to reduce the size of `pbench-index.py`, it is not clear how much this helps the situation.

The previous code was really about creating a generator pipeline, of sorts, for a given tar ball, and handing that generator to the es_index() method to send in bulk to Elasticsearch.

We might want to consider making a generator pattern for how tar balls to be indexed are gathered/found/produced, and pass that off to a consumer that deals with them the same way we do today.

We can then make the current method of how they are processed today, and add additional methods that perhaps send them to multiple workers for processing, etc.

Reply:
What if I just split the collecting tarballs function from the whole main module? and keep the other functions intact